vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Storvsc driver for use as a disk backend.
+
+pub mod test_helpers;
+
+use async_channel::Receiver;
+use async_channel::RecvError;
+use async_channel::Sender;
+use futures::FutureExt;
+use futures_concurrency::future::Race;
+use guestmem::AccessError;
+use guestmem::MemoryRead;
+use guestmem::ranges::PagedRange;
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use task_control::AsyncRun;
+use task_control::InspectTask;
+use task_control::StopTask;
+use task_control::TaskControl;
+use thiserror::Error;
+use tracing_helpers::ErrorValueExt;
+use vmbus_async::queue;
+use vmbus_async::queue::CompletionPacket;
+use vmbus_async::queue::ExternalDataError;
+use vmbus_async::queue::IncomingPacket;
+use vmbus_async::queue::OutgoingPacket;
+use vmbus_async::queue::PacketRef;
+use vmbus_async::queue::Queue;
+use vmbus_channel::RawAsyncChannel;
+use vmbus_ring::OutgoingPacketType;
+use vmbus_ring::PAGE_SIZE;
+use vmbus_ring::RingMem;
+use vmcore::vm_task::VmTaskDriverSource;
+use zerocopy::FromBytes;
+use zerocopy::Immutable;
+use zerocopy::IntoBytes;
+use zerocopy::KnownLayout;
+
+/// Storvsc to provide a backend for SCSI devices over VMBus.
+pub struct StorvscDriver<T: Send + Sync + RingMem> {
+    storvsc: TaskControl<StorvscState, Storvsc<T>>,
+    version: storvsp_protocol::ProtocolVersion,
+    driver_source: VmTaskDriverSource,
+    new_request_sender: Option<Sender<StorvscRequest>>,
+}
+
+/// Storvsc backend for SCSI devices.
+struct Storvsc<T: Send + Sync + RingMem> {
+    inner: StorvscInner,
+    version: storvsp_protocol::ProtocolVersion,
+    queue: Queue<T>,
+    num_sub_channels: Option<u16>,
+    has_negotiated: bool,
+}
+
+struct StorvscInner {
+    next_transaction_id: AtomicU64,
+    new_request_receiver: Receiver<StorvscRequest>,
+    transactions: Mutex<HashMap<u64, PendingOperation>>,
```

Comment:
I would use a `slab::Slab<PendingOperation>` instead of a `HashMap`. The slab will give you a `usize` key which you can use as the transaction ID.

Reply:
I see from the Slab docs that it can reuse keys for removed items. Could there be any issues with reused, non-sequential keys on the storvsp side?

Reply:
That would be a protocol violation--storvsp can only send a completion for a transaction once.

Reply:
So does slab still work for this if it can reuse keys, given that it can result in duplicate transaction IDs?